src/ucp/core/ucp_request.inl

```diff
@@ -553,7 +553,7 @@ ucp_request_recv_data_unpack(ucp_request_t *req, const void *data,
 static UCS_F_ALWAYS_INLINE ucs_status_t
 ucp_recv_desc_init(ucp_worker_h worker, void *data, size_t length,
                    int data_offset, unsigned am_flags, uint16_t hdr_len,
-                   uint16_t rdesc_flags, uint16_t priv_length,
+                   uint16_t rdesc_flags, int16_t priv_length,
```

Comment:
maybe rename "->priv_length" to "offset", and instead of using negative priv_length every AM receive will initialize the offset from uct_dest to rdesc, then ucp_recv_desc_release would be simpler

Reply:
can you please elaborate?
currently `priv_length` is used by tag offload, where we take extra space from headroom for the headers. Here it is opposite - we provide a pointer to the data (not header, and desc is made by `data_ptr - sizeof(rdesc)`), so we have to substruct the header size. If we change the semantic we will have to use negative priv_length for the tag offload case.

Reply:
i mean instead of using positive/negative length, save the offset from UCT desc, inside rdesc struct

Reply:
[Here](https://github.com/openucx/ucx/blob/master/src/ucp/core/ucp_request.inl#L593) we rely on `UCP_WORKER_HEADROOM_PRIV_SIZE ` to guess what is the correct offset for UCT. If we want to change it for rdesc field we will have to pass `priv_length` to rdesc_init anyway. And it will be negative either for HWTM or for AM. So the only thing which will be changed is
`uct_iface_release_desc(UCS_PTR_BYTE_OFFSET(rdesc, -rdesc->uct_desc_offset)`
 instead of
```
uct_iface_release_desc(UCS_PTR_BYTE_OFFSET(rdesc,
                                                   -(UCP_WORKER_HEADROOM_PRIV_SIZE -
                                                     rdesc->priv_length)));
```

Reply:
@yosefe?

Reply:
`uint16_t rdesc_flags, int priv_length,`
this function calculates HEADROOM+priv_length, make sure it's >=0, then safe it as rdesc->offset